### DISCUSSION TITLE
Proposal: @partial component renders

### DISCUSSION THREAD

**Discussion by danharrin** [livewire/livewire#7903] [3. Feature ideas]
When dealing with large Livewire components with lots of HTML, making updates can often just affect a small part of it. Currently, when using a single component, the entire HTML will render and get sent over the network for each update. Livewire will also spend time morphing the entire HTML.

The current way to avoid this is to break your Livewire components up, and nest them. This allows the isolated nested components to receive updates independently of the rest of the page. However, this is sometimes not convenient, if you need access to a single component instance / store of data in the backend, instead of multiple stores that need to stay in-sync.

This is where a second option can now come into play: partial Livewire component renders. The concept is that you are able to isolate parts of your Livewire component view so that you can decide if you want to render it independently of everything else. By default, partials are updated on each request, like normal. You must opt-in to a partial rendering indepdendently.

https://github.com/livewire/livewire/assets/41773797/c8f65810-14d5-4423-80ca-a3028f7bc04c

To define a partial, you must create a separate view from the main Livewire component view. This allows it to be compiled and evaluated independently, for performance reasons. A previous suggestion was that the partial views could be magically split up at compile time, but I imagine this would not work well in load balanced environments when different servers could have different sets of compiled views.

Once you've created the view, you must define another render method, specifically for that partial. Like so:

```php
public function renderFooBar(): View
{
    return view('livewire.partials.foo-bar');
}
```

In this case, the partial has the name `fooBar`, `foo_bar` or `foo-bar`, which is reflected in the name of the render method.

By default, the partial view has access to `$this`, and any public properties defined on the component. Any view data passed in `render()` is isolated.

To output the partial in the main Livewire component view:

```blade
@partial('fooBar') | @partial('foo_bar') | @partial('foo-bar')
```

For any requests that you wish to only render the partial for, you must call `renderPartialOnly()`:

```php
public function save(): void
{
    $this->renderPartialOnly('fooBar');
}
```

You could implement this in lifecycle hooks, to isolate the effects of a `wire:model.live` for example:

```php
public function updatedBaz(): void
{
    $this->renderPartialOnly('fooBar');
}
```

You could also use an alternative syntax using PHP attributes:

```php
#[RenderPartialOnly('fooBar')]
public function save(): void
{
    //
}
```

I have this mostly implemented and working as a plugin, but I believe it belongs in the Livewire core. Let me know what you think!

**Comment by pxlrbt** [comment]
I thought quite often about this, but didn't dare to try an implementation 😅 Partials really is a missing feature of Livewire and would really improve performance for complex Livewire apps.

**Comment by saade** [comment]
Cutting down on the amount of code sent over the wire certainly could make things load way faster and generally make the whole experience smoother. This optimization not only minimizes bandwidth usage but also prioritize performance for increasingly dynamic and interactive applications.

**Reply by Capevace** [reply]
Making it opt-in makes it super deliberate, I have lots of use cases for this! Feels like this really simplifies making large UI feel snappy, instead of having to rely on multiple components.

**Reply by alexandersix** [reply]
Agreed–I love the opt-in nature of this. I would love something like this, especially for the bandwidth usage reductions. Feels like I'd be able to make my interactive applications in a more deliberate way.

**Comment by ralphjsmit** [comment]
Fantastic feature, would be great if this could be added.

**Comment by binaryfire** [comment]
Our servers are in Europe but we have customers worldwide. Livewire apps can be a bit sluggish for users who are geographically far away from the server (eg. Australia -> Germany) because of the connection latency. Anything that reduces the amount of HTML that needs to be generated / sent over the wire will help and has my vote!

**Comment by pxlrbt** [comment]
I thought about the syntax for a while, and while I'm not sure if it's 100% possible, I think a more universal approach would be nice. Instead of `@partial` which requires to have a separate view file, it would be nice to have a `wire:partial` that you can use on any element and which indicates the target you want to replace. 

Then in the backend you can return something like `Partial::html('partial-id', 'your-html') ` or `Partial::view('partial-id', 'your-view-file')`. It would maybe even allow use to return multiple partials by using an array.

This would allow us to reuse Blade components as partials like the below example. Maybe we could even add a `Partial::component('partial-id', 'x-component', [...$data])` shortcut.

```html
<div id="my-component">
    <x-stats wire:partial="stats" :data="$data" />

    <main>
        @foreach($items as $key => $item)
            <x-item
                :data="$item"
                wire:partial="item-{{ $key }}"
                wire:click="updateRow({{ $key }})"
            />
        @endforeach
    </main>

    <x-summary wire:partial="summary" />
</div>
```

```php
<?php

class MyComponent
{
    public function updateRow($row)
    {
        // do something
        $rowData = $this->getRowData($row);

        return Partial::html(
            id: 'item-' . $row,
            html: Blade::render(
                '<x-item
                    :data="$data"
                    wire:partial="item-'.$row.'"
                    wire:click="updateRow('.$row.')"
                />',
                ['data' => $rowData]
            )
        );
    }    
}
```

**Reply by ralphjsmit** [reply]
That's an interesting approach! I personally think a mix of Dan's approach (not having a `Partial` class, but only a `renderPartialOnly()` method) and your approach (not needing to use `@partial`/other views, I would not like creating sub-views, unless technically necessary) would be nicest:

```blade
<div id="my-component">
    <div wire:partial="stats"><!-- --></div>

    <main>
        @foreach($items as $key => $item)
            <div wire:partial="row-{{ $item->getKey() }}">
               <!-- -->
            </div>
        @endforeach
    </main>

    <div wire:partial="summary"><!-- --!></div>
</div>
```

And then use it in PHP without any Partial class:

```php
class MyComponent
{
    public function updateRow($row)
    {
        // Do something
        $rowData = $this->getRowData($row);

        $this->renderPartialOnly("row-{$rowData->id}");
        // Any partial...
        $this->renderPartialOnly('stats');
    }    
}
```

**Reply by danharrin** [reply]
That would result in the entire view getting rendered to extract the partial, which isn't ideal. Hence why multiple files are used

**Reply by ralphjsmit** [reply]
I see, but it would improve DX imho and rendering compiled Blade is quite fast I think? The main issue is reducing the size of HTML sent over the wire and needing to be morphed.

**Reply by danharrin** [reply]
You'd need to run all the DB queries that are required to render the rest of the page - for example if you want to refresh the filters on a table without fetching any content

**Reply by ralphjsmit** [reply]
OK, agree that makes sense. In that case I agree separate Blade views are necessary, but if technically possible, a plain `wire:partial` on inline elements would be cool for smaller/custom components that aren't as heavy as Filament or where there's no effect in DB queries.

**Reply by pxlrbt** [reply]
> And then use it in PHP without any Partial class:

Yeah, I chose the Partial class for maximum flexibility. You can send whathever HTML you need and still reduce the queries.

**Comment by tylerwiegand** [comment]
i really love the idea, wish we could use the wire:partial="fooBar" syntax, but just to isolate part of the view to tell morphDom "only look here and in this subtree" instead of/in addition to having a separate partial file

**Comment by ju5t** [comment]
I like the idea. We have an app that has many large full page components with a lot of sections and moving parts. But I don't think we will use this. I'm not seeing the DX yet and I think this is going to add complexity to components that would actually benefit from slimming down/breaking up. I know this isn't always possible but eventually, that would be my preferred solution.

**Comment by aliqasemzadeh** [comment]
It looks like a good idea.
When we have non-countable records it might work hard but It works on some components.
I think we have to add some document about this in livewire documents.

**Comment by calebporzio** [comment]
Good input everyone. Here are my thoughts:

## Proposal

We add a simple directive called `@partial`. You can wrap any portion of your component's template with it like so:

```html
<div>
    <!-- ... -->
    
    @partial('filters')
        <!-- ... -->
    @endpartial

    <table>
        <!-- ... -->
    </table>
</div>
```

For normal renders, it's ignored.

If a method has the `#[Partial]` attribute attached, when it's called, it will only send the specified partial over the wire: 

```php
#[Partial('filters')]
public function updateFilters()
{
    // ...
}
```

When `updateFilters` is called on the component, the entire Blade template will be rendered, but only the contents of `@partial('filters')` will be sent to the browser. This also means only the partial contents will be "morphed".

### Pros

This approach improves a component's performance in two ways:
* A smaller payload sent to the browser
* Only a small portion of the template is "morphed"

This will result in faster and more reliable updates to small parts of a large component.

### Cons

The biggest con of this approach is: The entire component's template is rendered every time.

This means you might make database queries for variables passed into the view that aren't used inside the partial at all.

## An alternative approach

If we wanted to address the above con, we could provide an opt-in API for optimizing partials further (only rendering their contents instead of the entire component's template).

Here's how it would look.

```html
<div>
    <!-- ... -->
    
    @partial('filters')
        <!-- ... -->
    @endpartial

    <table>
        <!-- ... -->
    </table>
</div>
```

As you can see, the view part is unchanged, however, the backend part would include an additional render method dedicated to rendering the partial:

```php
#[Partial('filters')]
public function updateFilters()
{
    // ...
}

public function renderFilters($view)
{
    return $view->with([
        'some-scope',
    ]);
}
```

By including a `renderFilters` method, Livewire will now only render that partial with the provided scope, instead of calling the larger, slower, `render()` method of the component.

This solution would require Livewire to parse the Blade view and extract the partial portion of the template into a separate cached view file on the server.

Dan noted earlier that this could pose problems for servers under a load-balancer. It's true that this could potentially be a problem, but the conditions that would need to be met for this to even be a problem would be rare, and it wouldn't be hard to account for these scenarios and have a graceful fallback of just rendering the entire component like before.

## Wrapping up

In short, I think we should add at least a basic version of `@partial`, and if the value is worth the complexity, pursue the more advanced/capable version. The added complexity of option 2 won't be trivial and should be seriously considered as we way out the pros and cons.

There are still lots of micro decisions to make around naming and implementation details, but I wanted to put this out there as my "preferable" API for this feature so we can all discuss further.

**Reply by danharrin** [reply]
Sounds like it could be implemented with fragments instead of a new directive - https://laravel.com/docs/10.x/blade#rendering-blade-fragments

I would like to see a solution where you could use a fragment OR a partial, totally don't mean to come across rude here but rendering the entire template would make the feature effectively useless in my use case

`#[Fragment('foo')]` / `#[Partial('foo')]` with `renderPartial()`

**Reply by danharrin** [reply]
I do think that the "alternative approach", if we could get it to create the sub-view when `view:cache` is run, would allow us to solve it for load balanced environments where the views are cached on each server individually. That would be my preference if you don't want to implement two features, but I'm unsure how to hook into the compile process like that so you'd probably need to do the implementation

**Reply by calebporzio** [reply]
Yeah fragments are pretty much what we're doing, but I decided against using them for the following reasons:
* I think we would have to hijack the implementation of them for Livewire components only (and we've been bit by that in the past (remember `@stack` lol?))
* The term fragment is commonly used in front-end frameworks for "dom fragments" that allow you to render a node with multiple root siblings. Like `<></>` in JSX. And because Livewire/Alpine are also kind of front-end frameworks, I didn't want to use a term we might want to use later on for something else.

BUT it's basically the same feature in essence and I always try to use Laravel when I can to keep the amount of things people have to learn to a minimum.

As for the caching at the `view:cache` phase, I think we have to fix an unrelated bug with morph injection for pre-compiling, but this would be covered under that for sure so people could pre-compile their views.

But again, I honestly think it's not a big deal for load balanced environments because as long as the other server has rendered the component even a single time for a single user it will have cached partial for all future requests. So it's pretty edge-case-y, compared to other blade compiler/load-balancer problems

**Reply by danharrin** [reply]
I would like to suggest going down the route of the alternative approach then, avoiding the entire view from getting rendered would be a huge win IMO

**Reply by ju5t** [reply]
Instead of:

```php
#[Partial('filters')]
public function updateFilters()
{
    // ...
}
```

Would it be possible to do:

```php
#[Partial('filters')]
public array $filters = [];
```

Which would bind it to the correct scope as well? This way you don't need methods to figure out if something is inside a partial or not. I think this looks cleaner and people are already used to using attributes in Livewire. Perhaps we can also use it like this:

```php
#[Partial('filters', 'something-else')]
public array $filters = [];
```

This would scope the variable to two different partials and it should only render those two if something changes.

The attribute could target methods too.

**Reply by pxlrbt** [reply]
I agree with Dan, that the simple approach isn't really useful for most projects and the advanced approach should be the goal.

> This solution would require Livewire to parse the Blade view and extract the partial portion of the template into a separate cached view file on the server.

This could be prevented if users manually need to supply a view file/html/component for the partial, right? So less magic, but also less complexity on Livewire side. 

I'd also be careful with a `@partial` directive, as people might have registered that directive themselves?

~~What extra would it need to make Livewire work with `@fragment`? Would a unique key be enough for diffing on the frontend side? Then we could attempt a PR to Laravel entirely rely on fragments for the view part.~~ 

EDIT: Just tested and fragments still require all the vars, because they aren't extracted into separate view cache files. So I don't think they help here.

**Reply by samlev** [reply]
Another option that _might_ be easier to handle could be having the `@partial` directive being analogous to `@include`, but with the first parameter being the name:

```html
<div>
  <!-- -->
  @partial('foo', 'path.to.partial.view')
  <!-- -->
</div>
```

Or you can just pass the name `@partial('foo')`, and define a `renderFoo()` function.

This would mean that you wouldn't have to parse the blade file to pull partial templates out that might rely on variables from an outer scope.

**Comment by Zephni** [comment]
I know it would be a bit of a rework for how livewire handles re-renders, but maybe a @partial('name', 'method') component's $slot could be stored in memory client side?

So the {{ $slot }} of the @partial would be the the actual blade code so if we did:

```
@partial('name', 'method')
    <any blade code here>
@endpartial
```

Then the name, special render method, and `<any blade code here>` would get stored client side, and then when an update happens, the client side would be able to determine which partial belongs to which and can update that accordingly by sending a single request to that method and only getting what is necessary for that partial.

One important thing to note would be that the method would need to return just the updated variables with their original keys, so something like:

```
public function renderPartialName()
{
    return [
        'varName' => 'newVarValue',
       ... etc
    ];
}
```

This obviously seems to go outside the design philosophy a little, but it solves both the problem of having control over exactly over what gets re-rendered, and also absolute minimum overheads. I realise there would need to be some recognition from Livewire's main rendering to include the partials on first load, and then handle the storing of the $slot etc client side. I'm happy to look into it. One further thing to note would be that the blade "code" would be technically findable on the clients machine. But as long as user's are aware of this and not hard coding passwords into the @partial I can't personally see a problem with it.

**Comment by KhairulAzmi21** [comment]
Suddenly googling about livewire partial rendering before end of 2024 😅 and saw this !

**Comment by Dimimo** [comment]
Very good idea. In a sense, 'partials' already exist. At least in my experience. With [wire:replace](https://livewire.laravel.com/docs/wire-replace).

In a complex Livewire component (full page with Blade components), a checkbox refused to unset. All have `wire:key`'s and the method called by an action does change the DB value and updates the property. A page refresh shows the changes. Tried all, spend hours. Accidentally found out that `wire:replace` solved it. Although the docs don't mention such use case.  No idea why. I didn't take the time to study Livewire's under-the-hood javascript.

All loops (should!) have `wire:key`'s right? We could do something like `wire:replace('unique-key-id')`

[The partial code is here](https://github.com/Dimimo/pgbilliard/blob/5fcff31bb448be1239fb880086df89c922e09b0f/resources/views/components/schedule/position-dropdown.blade.php#L64C25-L73C33)

**Reply by pxlrbt** [reply]
All that `wire:replace` is, is to skip DOM morphing on that part. Otherwise there is not performance gain and you still need to render the whole component.

**Reply by samlev** [reply]
As @pxlrbt said - `wire:replace` is only a "performance improvement" in that it skips the morph step for an element. It's essentially the same "performance improvement" as `wire:ignore`, except the new DOM will need to be re-crawled by Alpine. The entire component is still rendered and passed back to browser.

A partial would be rendering only a tiny piece of the component, and passing that back instead of re-rendering everything.
